### PR TITLE
tests(github service): add github service tests

### PR DESCRIPTION
**Pull request by kishore03109**
## Problem

Github service behaviour was recently changed. The tests did not have good coverage of the added functionality.

Closes [insert issue #]

## Solution

Add new edge cases

### CODE REVIEW

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 1]
```diff
@@ -132,12 +151,88 @@ describe("Github Service", () => {
...
       })
       expect(mockAxiosInstance.put).toHaveBeenCalledWith(
-        endpoint,
+        folderEndpoint,
         params,
         authHeader
       )
     })
 
+    it("Creating a top level folder works correctly", async () => {
```
Comment (seaerchin) [inline src/services/db/__tests__/GitHubService.spec.ts:160]:
```suggestion
    it("should create a top level folder correctly", async () => {
```

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 2]
```diff
@@ -132,12 +151,88 @@ describe("Github Service", () => {
...
+        })
+      ).resolves.toMatchObject({
+        sha,
+      })
+      const resourceRoomParams = {
+        message: collectionYmlMessage,
+        content: Base64.encode(collectionYmlContent),
+        branch: params.branch,
+      }
+      const topLevelFolderEndpoint = `${folderParentEndpoint}/${topLevelDirectoryFileName}`
```
Comment (seaerchin) [inline src/services/db/__tests__/GitHubService.spec.ts:190]: nit: could we shift this to be grouped with the other arrange stuff (eg: initialising `resp`)

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 3]
```diff
@@ -132,12 +151,88 @@ describe("Github Service", () => {
...
+      expect(mockAxiosInstance.put).toHaveBeenCalledWith(
+        topLevelFolderEndpoint,
+        resourceRoomParams,
+        authHeader
+      )
+    })
+
+    it("Creating a resource category works correctly", async () => {
```
Comment (seaerchin) [inline src/services/db/__tests__/GitHubService.spec.ts:198]:
```suggestion
    it("should create a resource category correctly", async () => {
```

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 4]
```diff
@@ -132,12 +151,88 @@ describe("Github Service", () => {
...
+        })
+      ).resolves.toMatchObject({
+        sha,
+      })
+      const resourceRoomParams = {
+        message: resourceRoomMessage,
+        content: Base64.encode(indexHtmlContent),
+        branch: params.branch,
+      }
+      const resourceRoomFolderEndpoint = `${folderParentEndpoint}/${resourceCategoryFileName}`
```
Comment (seaerchin) [inline src/services/db/__tests__/GitHubService.spec.ts:228]: nit: shift to be w/ other arrange stuff

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 5]
```diff
@@ -185,11 +281,71 @@ describe("Github Service", () => {
...
       ).rejects.toThrowError(ConflictError)
       expect(mockAxiosInstance.put).toHaveBeenCalledWith(
-        endpoint,
+        folderEndpoint,
         params,
         authHeader
       )
     })
+
+    it("Create throws an error if parent directory is deleted", async () => {
```
Comment (seaerchin) [inline src/services/db/__tests__/GitHubService.spec.ts:290]:
```suggestion
    it("should throw an error if create is called when the parent directory is deleted", async () => {
```

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 6]
```diff
@@ -185,11 +281,71 @@ describe("Github Service", () => {
...
+      expect(mockAxiosInstance.get).toHaveBeenCalledWith(folderParentEndpoint, {
+        validateStatus,
+        headers: authHeader.headers,
+        params: {
+          ref: BRANCH_REF,
+        },
+      })
+    })
+
+    it("Create throws an error if a new sub directory is created while the parent directory is deleted", async () => {
```
Comment (seaerchin) [inline src/services/db/__tests__/GitHubService.spec.ts:310]:
```suggestion
    it("should throw an error when create is called if a new sub directory is created while the parent directory is deleted", async () => {
```

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 7]
```diff
@@ -185,11 +281,71 @@ describe("Github Service", () => {
...
+      mockAxiosInstance.get.mockImplementation(() => {
+        throw new Error()
+      })
+      await expect(
+        service.create(sessionData, {
+          content,
+          fileName: subDirectoryFileName,
+          directoryName: subDirectoryName,
+        })
+      ).rejects.toThrowError()
```
Comment (seaerchin) [inline src/services/db/__tests__/GitHubService.spec.ts:320]: nit: more specific error
```suggestion
      mockAxiosInstance.get.mockImplementation(() => {
        throw new NotFoundError()
      })
      await expect(
        service.create(sessionData, {
          content,
          fileName: subDirectoryFileName,
          directoryName: subDirectoryName,
        })
      ).rejects.toThrowError(NotFoundError)
```

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 8]
```diff
@@ -185,11 +281,71 @@ describe("Github Service", () => {
...
+      mockAxiosInstance.get.mockImplementation(() => {
+        throw new Error()
+      })
+      await expect(
+        service.create(sessionData, {
+          content,
+          fileName,
+          directoryName: `${resourceCategoryName}/_posts`,
+        })
+      ).rejects.toThrowError()
```
Comment (seaerchin) [inline src/services/db/__tests__/GitHubService.spec.ts:340]:
```suggestion
      mockAxiosInstance.get.mockImplementation(() => {
        throw new NotFoundError()
      })
      await expect(
        service.create(sessionData, {
          content,
          fileName,
          directoryName: `${resourceCategoryName}/_posts`,
        })
      ).rejects.toThrowError(NotFoundError)
```

**Review by seaerchin — APPROVED** [review]
mostly nits - i do feel that the error type one should be done so that it doesn't become a blindspot

**Review by seaerchin — APPROVED** [review]
mostly nits - i do feel that the error type one should be done so that it doesn't become a blindspot

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 9]
```diff
@@ -13,8 +13,11 @@ const {
   mockCurrentCommitSha,
   mockGithubSessionData,
   mockIsomerUserId,
-} = require("@fixtures/sessionData")
-const { GitHubService } = require("@services/db/GitHubService")
+} from "@fixtures/sessionData"
+import { GitHubService } from "@services/db/GitHubService"
+
+// using es6 gives some error
+const { Base64 } = require("js-base64")
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/services/db/__tests__/GitHubService.spec.ts:20]: nope, to be specific I got a 
```
Could not find a declaration file for module 'js-base64'. 'isomercms-backend/node_modules/js-base64/base64.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/js-base64` if it exists or add a new declaration (.d.ts) file containing `declare module 'js-base64';
```
the command `npm i --save-dev @types/js-base64` didnt have any effect on this error in my local machine

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 10]
```diff
@@ -28,7 +31,21 @@ describe("Github Service", () => {
...
   const content = "test-content"
+  const collectionYmlContent = `collections:
+  new-folder:
+    output: true
+    order: []`
+
+  const indexHtmlContent = `---
+  layout: resources-alt
+  title: some-folder
+  ---`
```
[the remark this answers is not in the document]
Comment (kishore03109) [reply src/services/db/__tests__/GitHubService.spec.ts]: good point! 
hmm the other fixtures dont use SCREAMING_SNAKE_CASE, feels inconsistent to have some have some dont have, you ok using them without?

**Comment by kishore03109** [comment]
@seaerchin thank you for the comment about the error typing, there was indeed a mistake in the code which didn't throw the error properly and lead to an internal server error. 
Have changed the behaviour just throw `NotFoundError` from `create` directly, and the corresponding test cases pass!

File: src/services/db/__tests__/GitHubService.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/services/db/__tests__/GitHubService.spec.ts:22]: oops forgot to tell you to not merge - can you check if it builds properly?
